File: lib/less/source-map-builder.js
module.exports = function (SourceMapOutput) {

    var SourceMapBuilder = function (options) {
        this.options = options;
    };

    SourceMapBuilder.prototype.toCSS = function(rootNode, options, imports) {
        var sourceMapOutput = new SourceMapOutput(
            {
                contentsIgnoredCharsMap: imports.contentsIgnoredChars,
                rootNode: rootNode,
                contentsMap: imports.contents,
                sourceMapFilename: this.options.sourceMapFilename,
                sourceMapURL: this.options.sourceMapURL,
                outputFilename: this.options.sourceMapOutputFilename,
                sourceMapBasepath: this.options.sourceMapBasepath,
                sourceMapRootpath: this.options.sourceMapRootpath,
                outputSourceFiles: this.options.outputSourceFiles,
                sourceMapGenerator: this.options.sourceMapGenerator,
                sourceMapFileInline: this.options.sourceMapFileInline
            });

        var css = sourceMapOutput.toCSS(options);
        this.sourceMap = sourceMapOutput.sourceMap;
        this.sourceMapURL = sourceMapOutput.sourceMapURL;
        if (this.options.sourceMapInputFilename) {
            this.sourceMapInputFilename = sourceMapOutput.normalizeFilename(this.options.sourceMapInputFilename);
        }
        return css;
    };

    SourceMapBuilder.prototype.getExternalSourceMap = function() {
        return this.sourceMap;
    };
    SourceMapBuilder.prototype.setExternalSourceMap = function(sourceMap) {
        this.sourceMap = sourceMap;
    };

    SourceMapBuilder.prototype.isInline = function() {
        return this.options.sourceMapFileInline;
    };
    SourceMapBuilder.prototype.getSourceMapURL = function() {
        return this.sourceMapURL;
    };
    SourceMapBuilder.prototype.getOutputFilename = function() {
        return this.options.sourceMapOutputFilename;
    };
    SourceMapBuilder.prototype.getInputFilename = function() {
        return this.sourceMapInputFilename;
    };

    return SourceMapBuilder;
};

File: packages/less/src/less/source_map_builder.go
package less

type SourceMapBuilder struct {
	options                   SourceMapBuilderOptions
	sourceMap                string
	sourceMapURL             string
	sourceMapInputFilename   string
}

type SourceMapBuilderOptions struct {
	SourceMapFilename              string
	SourceMapURL                   string
	SourceMapOutputFilename        string
	SourceMapInputFilename         string
	SourceMapBasepath              string
	SourceMapRootpath              string
	OutputSourceFiles              bool
	SourceMapGenerator             any
	SourceMapFileInline            bool
	DisableSourcemapAnnotation     bool
}

type SourceMapEnvironment interface {
	EncodeBase64(str string) string
}

type Imports struct {
	Contents              map[string]string
	ContentsIgnoredChars  map[string]int
}

func NewSourceMapBuilder(options SourceMapBuilderOptions) *SourceMapBuilder {
	return &SourceMapBuilder{
		options: options,
	}
}

func (smb *SourceMapBuilder) ToCSS(rootNode Node, options map[string]any, imports *Imports, environment SourceMapEnvironment) string {
	sourceMapOutputOptions := SourceMapOutputOptions{
		ContentsIgnoredCharsMap:        imports.ContentsIgnoredChars,
		RootNode:                       rootNode,
		ContentsMap:                    imports.Contents,
		SourceMapFilename:              smb.options.SourceMapFilename,
		SourceMapURL:                   smb.options.SourceMapURL,
		OutputFilename:                 smb.options.SourceMapOutputFilename,
		SourceMapBasepath:              smb.options.SourceMapBasepath,
		SourceMapRootpath:              smb.options.SourceMapRootpath,
		OutputSourceFiles:              smb.options.OutputSourceFiles,
		SourceMapGeneratorConstructor:  func() SourceMapGenerator { 
			// This should be injected from the caller in real usage
			// For now, we'll use a default implementation
			return &defaultSourceMapGenerator{
				mappings: make([]SourceMapMapping, 0),
				sourceContents: make(map[string]string),
			}
		},
	}

	sourceMapOutput := NewSourceMapOutput(sourceMapOutputOptions)
	css := sourceMapOutput.ToCSS(options)
	smb.sourceMap = sourceMapOutput.SourceMap
	smb.sourceMapURL = sourceMapOutput.sourceMapURL
	
	if smb.options.SourceMapInputFilename != "" {
		smb.sourceMapInputFilename = sourceMapOutput.NormalizeFilename(smb.options.SourceMapInputFilename)
	}
	
	if smb.options.SourceMapBasepath != "" && smb.sourceMapURL != "" {
		smb.sourceMapURL = sourceMapOutput.RemoveBasepath(smb.sourceMapURL)
	}
	
	return css + smb.getCSSAppendage(environment)
}

func (smb *SourceMapBuilder) getCSSAppendage(environment SourceMapEnvironment) string {
	sourceMapURL := smb.sourceMapURL
	
	if smb.options.SourceMapFileInline {
		if smb.sourceMap == "" {
			return ""
		}
		sourceMapURL = "data:application/json;base64," + environment.EncodeBase64(smb.sourceMap)
	}

	if smb.options.DisableSourcemapAnnotation {
		return ""
	}

	if sourceMapURL != "" {
		return "/*# sourceMappingURL=" + sourceMapURL + " */"
	}
	return ""
}

func (smb *SourceMapBuilder) GetExternalSourceMap() string {
	return smb.sourceMap
}

func (smb *SourceMapBuilder) SetExternalSourceMap(sourceMap string) {
	smb.sourceMap = sourceMap
}

func (smb *SourceMapBuilder) IsInline() bool {
	return smb.options.SourceMapFileInline
}

func (smb *SourceMapBuilder) GetSourceMapURL() string {
	return smb.sourceMapURL
}

func (smb *SourceMapBuilder) GetOutputFilename() string {
	return smb.options.SourceMapOutputFilename
}

func (smb *SourceMapBuilder) GetInputFilename() string {
	return smb.sourceMapInputFilename
}

// defaultSourceMapGenerator provides a default implementation for testing
type defaultSourceMapGenerator struct {
	mappings       []SourceMapMapping
	sourceContents map[string]string
}

func (d *defaultSourceMapGenerator) AddMapping(mapping SourceMapMapping) {
	d.mappings = append(d.mappings, mapping)
}

func (d *defaultSourceMapGenerator) SetSourceContent(source, content string) {
	d.sourceContents[source] = content
}

func (d *defaultSourceMapGenerator) ToJSON() map[string]any {
	return map[string]any{
		"version":  3,
		"sources":  []string{},
		"mappings": "",
	}
}